### ISSUE TITLE
SC2155: A hidden case of "masking" return values

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#2770]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2155
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

set -Eeuo pipefail

function get_x() {
  local x=$(cmd-that-fails "$1" | cmd-that-fails-moar)
  echo "$x"
}

x=$(get_x "abc")
```

#### Here's what shellcheck currently says:

```console
$ shellcheck myscript
No issues detected!

$
```

#### Here's what I wanted or expected to see:

```console
$ shellcheck myscript
 
Line 6-7:
  echo "$x"
^-- [SC2155](https://www.shellcheck.net/wiki/SC2155) (warning): Assigning a command output to $x and then returning the result via `echo $x` will not propagate a potential exit code. Instead, replace with the command

$
```

**Comment by TinCanTech** [comment]
@stdedos be careful, you appear to have stepped on **two** dragons..

* https://mywiki.wooledge.org/BashFAQ/105
* https://en.wikipedia.org/wiki/Semipredicate_problem

**Comment by stdedos** [comment]
Well ... let's leave the https://mywiki.wooledge.org/BashFAQ/105 out of the discussion ...

"I know" of the flaky-ness of `set -e`. My attitude with `bash` is "WOMM" 😅 (but this is not running on just my machine)
Or _don't use it altogether_ ...

-----

Would you elaborate on how the https://en.wikipedia.org/wiki/Semipredicate_problem applies? I am struggling to apply it here 😕 

My assumption is that the developer seems to be falling into a variant of a logical fallacy akin to the [Useless Use of Cat (UUOC)](http://en.wikipedia.org/wiki/Cat_(Unix)#Useless_use_of_cat) anti-pattern. I'd say that the situation here is "well-enough" defined to make a suggestion (it hangs on the `set -e` and `$(get_x)` though 😕)

Ofc koalaman might have a different opinion about it wrt popularity of such error, given his corpus of scripts.

**Comment by TinCanTech** [comment]
Your code does zero error checking, therefore you are reliant on `set -e`.

**Comment by TinCanTech** [comment]
As for the other problem: Your code has returned no error, as instructed.

`shellcheck` is faced with the same problem, there is nothing wrong.

There are ways that you can do the error checking for yourself but you rely on `set -e`.
